### DISCUSSION TITLE
Optimization of included JS files

### DISCUSSION THREAD

**Discussion by firai** [zauberzeug/nicegui#1259] [Ideas / Feature Requests]
I noticed that the size of the NiceGui package, not including dependencies, increased from 13 MB to 49 MB between 1.2.x to 1.3.x. I understand that this is due to #658, which copied in the full JS packages rather than the minified versions. Upon further inspection, the copied packages (at least for mermaid, plotly and aggrid) appear to _also_ contain the minified versions. If the full packages files are being used now, can/should the minified files be removed to eliminate duplication?

My JS knowledge is definitely rudimentary, so I apologize if what I'm suggesting above doesn't make sense given how the packages are set up.

**Comment by falkoschindler** [comment]
Very interesting observation, @firai!

Honestly, I didn't check which files are provided for every single dependency. Sure, in case we have both versions and use only one, we could remove the other one. But we also wanted to provide source maps for easier debugging. As far as I know, you need the non-minified file for source maps to work. So we ultimately need to provide *.js, *.min.js _and_ *.js.map, growing the package size even more. Or is *.min.js and *.js.map enough?

**Comment by firai** [comment]
Thanks for your response, @falkoschindler! Bard and Bing seem to tell me that you could still get somewhat readable debugging information without the full package, but I don't really debug JS/TS. @dclause, do you know the answer to @falkoschindler's question? It seems like Mermaid stopped building source maps into the npm/CDN package since 10.2 though.

Looking at Mermaid as an example (since that's the biggest package), it seems to me that ~~4~~5 versions of the package are actually being shipped with NiceGui right now:
- ESM full (`mermaid.esm.mjs` and its tree of files)
- ESM minified (`mermaid.esm.min.mjs` and its tree of files)
- ESM core bundle (`mermaid.core.mjs` and its tree of files)
- UMD full (`mermaid.js`)
- UMD minified (`mermaid.min.js`)

If we need the full packages for debugging and if we have switched to ESM, maybe we can at least get rid of the UMD packages? Ideal scenario would be getting down to just one version.

EDIT: Just thinking out loud, could we somehow point to the CDNs for the full versions for debugging? Might be unlikely that debugging is being done on a machine with no network access? And _if_ that's desired by the user, maybe we can tell the user (or provide a script) to download the full versions as needed, instead of distributing them to everyone?

**Reply by rodja** [reply]
> could we somehow point to the CDNs for the full versions for debugging?

Great idea! I'm really in favour of this approach.

**Reply by falkoschindler** [reply]
Me too!
Clarifying what is actually needed for debugging would be helpful though.

**Reply by dclause** [reply]
Sourcemap files maps the minified code to the original code. What that means is you don't need both the original and the sourcemap code to be delivered in front: either you serve the original code, or the minified one+sourcemap files.

Currently, we don't have a toggle way to serve minified / unminified code or debug sourcemap files. 
Also, the some of the libraries we are using are now ESM, hence spread over multiple files (and larger codebase), enabling "tree shaking" or at least, "include only what's needed". 
#658 is based on a web approach, that we are not used in software development, but mixed with the contraint of being a software product first. In other word: on the web, you don't ship dependencies with the project. A composer.json (PHP world) or a package.json (JS world) let's your package dependency download what's needed for you. Just like what poetry does (python world). But we do not want to ship with that need for a JS package dependency here, so you don't need a whole node/npm stack to use this python project. Therefore, the dependencies ships with the project.
It is hard to download an ESM library manually - that is not made to be done - you always go through the packagist (npm). Also in the web, we don't care the size of a library dependency itself ( in the sense of what's downloaded by the packagist), because it sits on the server and is not served. We only care to optimize what we serve to the client and how we do it (in an optimal way hopefully).

Using a CDN to download / retrieve the assets would be the easiest. It would solve some of the inclusion / issues we have. Remove workload from your python server, reduce the project size, eliminate the difficulty to retrieve the files (npm.py script), etc... But that means your client machine would need constant internet access to work. Is that something we want ? From my understanding, the porject was primarily done to serve at building easy configuration UI for robots that don't have internet access. Meaning you connect to there locally created network and you are served with this "offline" UI. 

If we decide that the UI can only be used from an internet accessible client, than we could go with the CDN option (providing also all libraries are compatible with that?)

I do suppose that removing the none minified version of the dependencies (when a minified one is available) would be the best. We could also imagine adding a new python dependency (dev only dependency) to expand our script and force minification of all JS files when the upgrade script download them in the project. That way, we would only serve minified JS.

NOTE: there is no modern framework anymore that does not come with a build phase where gulp, webpack, nitro, vite or other tooling would handle "babelification", minification and bundling, etc.. We have nothing of that here and that makes it hard to use modern JS library on our stack.
NOTE2: we can also decide to not support ESM or any library that could not be downloaded minified "old style", hence older version of plotly, mermaid, highcharts, threeJS, etc...

Too much words in my answer ! Does it help understanding the situation ?

**Reply by dclause** [reply]
Another note sorry: we don't control the sourcemap (at least that I know). They are autodownloaded by the browser developper tools are you open them. Until you open the tools, no sourcemap is downloaded by the browser, and when you do, the browser does so automatically.
What you discribe may be possible (I don't know everything for sure), but I don't know how to do so, meaning how to serve the files from some place (locally on the server) and indicate the browser to get the sourcemap from another place (a CDN, that might be different for every libraries).

**Reply by falkoschindler** [reply]
Thanks for your input, @dclause! 😀

Actually, there is an endpoint in nicegui.py that handles *.js and *.js.map requests:
https://github.com/zauberzeug/nicegui/blob/196ad1855923f3eab6a59498087d3c57aed0081c/nicegui/nicegui.py#L45-L56
We could probably forward source map requests to an online CDN pretty easily. It will only be required when dev tools are opened, so performance is not a huge issue. We just need to find the right URL for the requested source map. Maybe a look into npm.json helps. Apart from that we would only need minified JS files, and probably only ESM _or_ UMD.

All in all this sounds very promising. Besides reducing the PyPI package size, we'd also reduce the size of packaged NiceGUI apps.

**Reply by dclause** [reply]
Good idea indeed. That would require that the server has internet access rather than the client for debug. But still, I guess that is manageable. 
How about minifying the js ourself when downloading the library?

**Reply by dclause** [reply]
Maybe adding a need for JS stack on dev is not a bad idea ? That way we could have a build phase on dev when preparing a release.
Can we even imagine not at all having the libraries in the repo but building them on GitHub in the release pipeline? That way we would not maintain a weird npm.py script ?

**Reply by firai** [reply]
With regards to separating the minified Mermaid ESM package from the full, it seems like we might be able to write a Python script at download time (NiceGui dependency upgrade time) that checks the imports (via regex/search rather than actually parsing JS) to categorize them. For example, for Mermaid v10.2.4 that's currently in the NiceGui repo, `mermaid.esm.min.mjs` points to `mermaid-be6aa4a6.js`, as opposed to `mermaid-aad43469.js`, used by [the core bundle](https://www.sidharth.dev/posts/shrinking-mermaid/) `mermaid.core.mjs`, or `mermaid-bf016224.js`, used by the full package `mermaid.esm.mjs`. All other component files seem to import from one of these 3 files, indicating the tree that the file belongs to.

I haven't looked into three.js or Highcharts, the other package(s) supposedly migrated to ESM, to see how those work, but those may not matter as much since the sizes are smaller. Plotly and aggrid still seem to use single-file UMD builds, so picking the mini versions should be easy.

---
### ESM full

mermaid.esm.mjs
```js
import { aP } from "./mermaid-bf016224.js";
```

classDiagram-50a13b48.js
```js
import { c as getConfig, l as log, j as select, k as configureSvgSize } from "./mermaid-bf016224.js";
```
---
### ESM mini

mermaid.esm.min.mjs
```js
import { aP as f } from "./mermaid-be6aa4a6.js";
```

classDiagram-440b88cd.js
```js
import { c as S, l as d, j as m, k as X } from "./mermaid-be6aa4a6.js";
```
---
### ESM core

mermaid.core.mjs
```js
import { K } from "./mermaid-aad43469.js";
```

classDiagram-d26c05e1.js
```js
import { c as getConfig, l as log, i as configureSvgSize } from "./mermaid-aad43469.js";
```

**Reply by falkoschindler** [reply]
Super exciting thoughts!

@dclause I think if a user really wants to debug into JS code, the requirement of an internet connection is fine.
And yes, I could imagine to have the JS stack as part of the dev dependencies.
But what about users that want to checkout and use the current main branch? Are they "devs", i.e. would need the JS stack and build the dependencies themselves? This sounds not very convenient. Or should we keep prebuilt JS dependencies in the repo for such a use case?

@firai Even though I couldn't allow all the details, it sounds like a reasonable and very powerful approach for reducing the dependencies to what is actually needed. Do you think this method would be applicable to future versions of Mermaid etc., or will we need to make adjustments to, e.g., the regex when updating Mermaid?

**Reply by firai** [reply]
@falkoschindler I was originally suggesting to traverse the Mermaid ESM module tree bottom up while only reading the first few lines of each file, but it turns out there are a couple of modules (`array` and `init`) that don't point back to the head. The more reliable way is therefore to traverse the tree top down and read the files in their entirety. See working code at https://gist.github.com/firai/4d2c981869a067a12ce08c2e355ffc18 to find all minified ESM files for both Mermaid 10.3.0 and 10.2.4 (the version currently copied into NiceGui). Takes about 10 seconds to find all of the files for each of the builds. The lists seem to be correct as:
- All modules are included.
- Included modules point back to the version of the main `mermaid-xxx.js` file that mermaid.esm.min.mjs refers to (except for `array` and `init` mentioned above, which don't import from the main module for some reason).
- The versions of `array` and `init` included are clearly the minified versions when you compare them against the other versions in the folder.

~~Note that I wrote this code based on Mermaid, since that's the biggest package in NiceGui right now, but it may work for other packages as well as long as the right root directories and entry points are chosen. As long as Mermaid uses the current directory structure (all of their js files are in the dist root directory), the code should work. If they change the directory structure, the code probably still works, as long as they prepend the imported paths with `./`, otherwise minor adjustments to the regex may be required.~~ To foolproof and future-proof this code, we could use `urllib.parse` to resolve the relative urls in the imports.

**Reply by falkoschindler** [reply]
@firai This sounds very promising. Would you like to create a PR to provide optimized Mermaid dependencies as a proof-of-concept? Or should we tackle _all_ dependencies at once?

**Reply by firai** [reply]
Is there a conclusion about how you want to deal with source maps for packages that don't provide them as build artifacts, such as Mermaid? Can we leave it as an upstream issue and not provide source maps if upstream doesn't provide them? Or do you want to minify the scripts ourselves?
(By the way, webpack is a bit beyond my realm of knowledge. I can try to take a stab at implementing the optimization for Mermaid without providing source maps, but someone may have to take this over if source maps need to be provided for all packages in this project.)

**Reply by falkoschindler** [reply]
Let's first focus on removing unneeded files. We can tackle missing source maps in a separate, later PR. This approach avoids cluttering a single PR and eases the review process.
